### ISSUE TITLE
Multiple Celebration Promotion emails can be sent

### ISSUE THREAD

**Issue by joemasilotti** [joemasilotti/railsdevs.com#617]
## Describe the bug

Under certain scenarios businesses are able to additional infinite Celebration Promotion emails to a developer.

## To reproduce

1. Reset your local database seeds via `bin/rails db:reset`
1. Sign in as `business@example.com`
1. Navigate to [the conversation with Tobias](http://localhost:3000/conversations/1)
1. Send a message

## Expected behavior

A second Celebration Promotion email is sent to the developer.

## Actual behavior

The second Celebration Promotion email is not sent to the developer.

## Additional context

The Celebration Promotion email should only be sent when:

1. The developer has responded to a conversation exactly once.
2. The new message is sent from the developer.

The existing code does not handle the second scenario. `app/models/concerns/messages/notifications.rb:36` needs to be adjusted.

**Comment by sarvaiyanidhi** [comment]
Hi @joemasilotti Can you please assign me this issue?

**Comment by joemasilotti** [comment]
Sure thing! All yours @sarvaiyanidhi.

**Comment by sarvaiyanidhi** [comment]
@joemasilotti I have gone through the existing flow and checked locally. I see that right now celebration promotion mail triggers only on first reply which I understand is due to below code called from notification


```
  def first_reply?(developer)
    messages.where(sender: developer).one?
  end

```

Now as per the requirement we want to send celebration promotion mail whenever there is any new message from developer which I could get locally working by changing above code to this

```
  def developer_reply?(developer)
    messages.where(sender: developer).any? && messages.last.sender_type == "Developer"
  end
```

By changing above code, it is now sending celebration promotion mails whenever there is any message from developer but I have few queries on the same before I proceed with implementation and create PR

- Here now if developer sends two messages one after the other in that case we will have two celebration promotion emails sent to them.. Is it correct as mail gets trigger on save of new message?
- I am not sure if there is any check right now wherein if developer has submitted form sent in celebration mail after receiving first mail, but after that if he sends again new message so code will again trigger new celebration promotion email although developer has already submitted form.. is it valid?

**Comment by joemasilotti** [comment]
Thanks for diving in! Here's the current flow:
1. Business messages a developer
2. Developer replies -> one celebration email is queued
3. Business replies -> **a _second_ celebration email is queued**
3. Business replies again -> **a _third_ celebration email is queued**4. 

And here's the expected, correct flow:
1. Business messages a developer
2. Developer replies -> one celebration email is queued
3. Business replies -> _no_ celebration email is queued
3. Business replies again -> _no_ celebration email is queued 

Said another way, only one should be triggered per conversation. No matter how many times either participant replies. Does that make more sense than what I originally wrote?

**Comment by sarvaiyanidhi** [comment]
Thanks @joemasilotti .. It is clear now. My bad I understood initially other way round.

Will look into issue and fix the bug further.

**Comment by joemasilotti** [comment]
Great! Sorry for the unclear description and thanks for diving in.

**Comment by joemasilotti** [comment]
Closed via #637.
